File: core/api/site_handlers.ts
import Elysia, { t } from "elysia";
import { DB } from "../store/db";
import { getAllSites, createSite } from "../store/sites";

export function siteHandlers(db: DB) {
    return new Elysia()
        .get("/sites", () => {
            try {
                const sites = getAllSites(db);
                return sites;
            } catch (err: any) {
                return new Response(JSON.stringify({ error: err.message }), { status: 500, headers: { "Content-Type": "application/json" } });
            }
        })
        .post("/sites", ({ body }) => {
            try {
                const { id, url, strategy, interval_seconds, enabled } = body;
                createSite(db, {
                    id,
                    url,
                    strategy: strategy || "text",
                    interval_seconds: interval_seconds || 600,
                    enabled: enabled !== undefined ? enabled : 1
                });
                return { status: "created", id };
            } catch (err: any) {
                return new Response(JSON.stringify({ error: err.message }), { status: 500, headers: { "Content-Type": "application/json" } });
            }
        }, {
            body: t.Object({
                id: t.String(),
                url: t.String(),
                strategy: t.Optional(t.String()),
                interval_seconds: t.Optional(t.Number()),
                enabled: t.Optional(t.Number())
            })
        });
}

File: core-golang/api/site_handlers.go
package api

import (
	"encoding/json"
	"fmt"
	"net/http"
	"time"
	"vigil-core/logger"
	"vigil-core/store"
)

// handleGetSites returns the list of all configured sites.
// GET /sites
func (s *Server) handleGetSites(w http.ResponseWriter, r *http.Request) {
	log := logger.Get()

	sites, err := s.db.GetAllSites()
	if err != nil {
		log.Error(fmt.Sprintf("Error getting sites: %v", err))
		writeError(w, http.StatusInternalServerError, "failed to get sites")
		return
	}

	if sites == nil {
		sites = []store.Site{}
	}
	writeJSON(w, http.StatusOK, sites)
}

// handlePostSites adds a new site to the database.
// POST /sites
func (s *Server) handlePostSites(w http.ResponseWriter, r *http.Request) {
	log := logger.Get()

	var site store.Site
	if err := json.NewDecoder(r.Body).Decode(&site); err != nil {
		writeError(w, http.StatusBadRequest, "invalid JSON body")
		return
	}

	if site.ID == "" || site.URL == "" {
		writeError(w, http.StatusBadRequest, "id and url are required")
		return
	}

	// Always enable by default when created via UI
	site.Enabled = true
	site.CreatedAt = time.Now().UTC()

	// Upsert the site
	if err := s.db.UpsertSite(&site); err != nil {
		log.Error(fmt.Sprintf("Error saving site: %v", err))
		writeError(w, http.StatusInternalServerError, "failed to save site")
		return
	}

	log.Info(fmt.Sprintf("New site added: %s", site.ID))

	writeJSON(w, http.StatusCreated, map[string]any{
		"status":  "ok",
		"message": "site created successfully",
		"site":    site,
	})
}
